### DISCUSSION TITLE
Which usage does the `ReferenceProps` mean?

### DISCUSSION THREAD

**Discussion by LoTwT** [vue-macros/vue-macros#558] [General]
Here is the definition of the `Props` type which is returned by `analyzeSFC`.
![image](https://github.com/vue-macros/vue-macros/assets/50254496/80be466e-df21-42e0-aa16-ae3ecfc9523e)

I know three usages of the `defineProps` as following:
1. defineProps(["msg"])
2. defineProps({ msg: String })
3. defineProps<{ msg: string }>()

I think the `TSProps` is 3 and the `ObjectProps` may be 2 .

## Questions

Which usage does the `ReferenceProps` type mean?
Does 1 also mean the `ObjectProps`?

**Comment by sxzz** [comment]
`Array` is also essentially an object, so cases 1 and 2 are `ObjectProps`. `ReferenceProps` stands for `defineProps(variable)` and applies to all non-literal expressions.
- `defineProps(variable)`
- `defineProps(generator())`

However, they're not implemented yet. Only TS-type declaration is supported.

**Reply by LoTwT** [reply]
Except for the feature `hoistStatic` of `vue-macros`, does `vue` support `ReferenceProps` itself currently?
I'd like to support `ObjectProps` and `ReferenceProps`. Are there some caveats I should pay more attention to?

**Reply by sxzz** [reply]
Yes. Vue core supported. For example:
```ts
import { buttonProps } from './foo'
defineProps(buttonProps)
```
